New Function Pointer Syntax

333fred opened the pull request:
This updates function pointers to use the new syntax decided on in LDM. This is syntax-level changes only: the updated calling convention _binding_ will come in the next PR.

Fixes https://github.com/dotnet/roslyn/issues/44312.

alrz commented on src/Compilers/CSharp/Portable/Syntax/Syntax.xml:
```diff
@@ -260,6 +271,59 @@
+      </PropertyComment>
+    </Field>
+  </Node>
+  <Node Name="FunctionPointerUnmanagedCallingConventionSyntaxList" Base="CSharpSyntaxNode">
```
by convention
```suggestion
  <Node Name="FunctionPointerUnmanagedCallingConventionListSyntax" Base="CSharpSyntaxNode">
```

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Generated/CSharp.Generated.g4:
```diff
@@ -348,7 +348,28 @@ array_rank_specifier
+  ;
+
+function_pointer_parameter
+  : attribute_list* modifier* type
```
i just realized something.  we don't allow identifiers for these parameters.  However, does this mean that if i have a funcptr, i can't used named-parameters with it?

333fred replied:
Correct. There's no place to encode parameter names in metadata.

333fred commented:
@dotnet/roslyn-compiler @AlekseyTs, this is now ready for review.

@dotnet/roslyn-ide, this includes syntax changes to function pointer types, and formatter/intellisense context fixes to ensure the existing tests still pass. We'll still need to make intellisense changes for the calling convention location at a later point.

333fred commented on src/Workspaces/CSharpTest/Formatting/FormattingTests_FunctionPointers.cs:
```diff
@@ -31,19 +30,90 @@ unsafe class C
+            await AssertFormatAsync(expected, content);
+        }
+
+        [Fact(Skip = "PROTOTYPE(func-ptr): Even though the rule is returning put a space, it's getting ignored, resulting in delegate*invalid <")]
```
I'm assuming that this is because `invalid` is parsed as trivia on a missing `managed` or `unmanaged` token? Those two are the only valid token types for the spot.

jasonmalinowski replied:
The "invalid" token isn't being parsed into the calling convention position? It looks based on the formatting change that's what you intended, right?

333fred replied:
It is parsed into the calling convention position, it's parsed as trivia on a missing `managed` token.

---
In reply to: [451061448](https://github.com/dotnet/roslyn/pull/45684#discussion_r451061448) [](ancestors = 451061448)

jasonmalinowski replied:
So I guess if the formatter's rule is we don't touch things around missing stuff, then maybe the behavior you're seeing is correct. Once the user has typed _correct_ code, it'll format correctly; if the user put their code into a broken state while they're doing other editing, not touching is fine for that small window.

jasonmalinowski commented on src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -211,6 +297,9 @@ static Microsoft.CodeAnalysis.CSharp.SyntaxFactory.WithExpression(Microsoft.Code
 *REMOVED*Microsoft.CodeAnalysis.CSharp.Syntax.ObjectCreationExpressionSyntax.ArgumentList.get -> Microsoft.CodeAnalysis.CSharp.Syntax.ArgumentListSyntax
 *REMOVED*Microsoft.CodeAnalysis.CSharp.Syntax.ObjectCreationExpressionSyntax.Initializer.get -> Microsoft.CodeAnalysis.CSharp.Syntax.InitializerExpressionSyntax
 *REMOVED*Microsoft.CodeAnalysis.CSharp.Syntax.ObjectCreationExpressionSyntax.NewKeyword.get -> Microsoft.CodeAnalysis.SyntaxToken
+*REMOVED*Microsoft.CodeAnalysis.CSharp.Syntax.ParameterSyntax.AttributeLists.get -> Microsoft.CodeAnalysis.SyntaxList<Microsoft.CodeAnalysis.CSharp.Syntax.AttributeListSyntax>
```
Doing this and _also_ updating the PublicAPI.Shipped.txt file? I'm a bit confused.

333fred replied:
I'm not sure what the quickfix did here, I'll double check.

jasonmalinowski commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:
```diff
@@ -1086,9 +1098,9 @@ public static bool IsDefaultExpressionContext(this SyntaxTree syntaxTree, int po
 #if !CODE_STYLE
             if (token.IsKind(SyntaxKind.CommaToken) &&
-                token.Parent.IsKind(SyntaxKindEx.FunctionPointerType, out FunctionPointerTypeSyntax funcPtrType))
+                token.Parent.IsKind(SyntaxKindEx.FunctionPointerParameterList, out FunctionPointerParameterList funcPtrParamList))
```
Is this unifiable now with the code above in some way now that the structure is similar?

333fred replied:
No, `FunctionPointerParameterList` and `ParameterListSyntax` don't share a common base type.

---
In reply to: [451057978](https://github.com/dotnet/roslyn/pull/45684#discussion_r451057978) [](ancestors = 451057978)

jasonmalinowski commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:
```diff
@@ -961,15 +961,27 @@ public static bool IsDefaultExpressionContext(this SyntaxTree syntaxTree, int po
                 case SyntaxKind.LessThanToken:
                 case SyntaxKind.CommaToken:
-                    return token.Parent is FunctionPointerTypeSyntax;
+                    return token.Parent is FunctionPointerParameterList;
```
Looks like you can delete the reference to https://github.com/dotnet/roslyn/issues/39865 now on line 959.

jasonmalinowski commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/SpacingFormattingRule.cs:
```diff
@@ -330,23 +330,88 @@ public override AbstractFormattingRule WithOptions(AnalyzerConfigOptions options
+                if (previousKind == SyntaxKind.AsteriskToken && currentParentKind == SyntaxKindEx.FunctionPointerCallingConvention)
                 {
-                    return CreateAdjustSpacesOperation(1, AdjustSpacesOption.ForceSpacesIfOnSingleLine);
+                    switch (currentKind)
+                    {
+                        case SyntaxKind.IdentifierToken:
+                        case SyntaxKindEx.ManagedKeyword:
+                        case SyntaxKindEx.UnmanagedKeyword:
+                            return CreateAdjustSpacesOperation(1, AdjustSpacesOption.ForceSpacesIfOnSingleLine);
+                    }
+                }
```
If we just checked that the currentToken is the FunctionPointerCallingConvention's ManagedSpecifier then we can remove this block, right? And we're having to do this because of the funkiness around not being able to reference new syntax node types in our code?

333fred replied:
Yeah, we could accomplish it that way.

---
In reply to: [451066787](https://github.com/dotnet/roslyn/pull/45684#discussion_r451066787) [](ancestors = 451066787)

jasonmalinowski replied:
@333fred If you could either do that work or just file a bug to do it, I think I'd feel better since this then is a bit more flexible around precisely what we might get out of the parser.

333fred replied:
Filed https://github.com/dotnet/roslyn/issues/45897.

jasonmalinowski commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:
```diff
@@ -961,15 +961,27 @@ public static bool IsDefaultExpressionContext(this SyntaxTree syntaxTree, int po
 #endif
         }
 
+        public static bool IsFunctionPointerParameterOrReturnType(this SyntaxToken syntaxToken)
```
Was this being used elsewhere or can it be inlined? (Sorry for the dumb question but it seems text search in my browser isn't being consistent so...)

333fred replied:
I thought it was being used elsewhere, but I apparently removed that need.

---
In reply to: [451068913](https://github.com/dotnet/roslyn/pull/45684#discussion_r451068913) [](ancestors = 451068913)

jasonmalinowski commented on src/Features/CSharp/Portable/Completion/KeywordRecommenders/ReadOnlyKeywordRecommender.cs:
```diff
@@ -43,7 +43,7 @@ protected override bool IsValidContext(int position, CSharpSyntaxContext context
             => context.TargetToken.IsKind(SyntaxKind.RefKeyword) &&
                (context.TargetToken.Parent.IsKind(SyntaxKind.RefType) ||
-                (context.TargetToken.Parent.IsKind(SyntaxKind.Parameter) && context.IsFunctionPointerTypeArgumentContext));
+                (context.TargetToken.Parent.IsKind(SyntaxKind.FunctionPointerParameter) && context.IsFunctionPointerTypeArgumentContext));
```
Deleting the IsKind() check doesn't work anymore? I would have assumed context.IsFunctionPointerTypeArgumentContext has already figured this out accurately enough and we don't have to do extra guesses.

333fred replied:
I think we can remove this. I'll commit and we'll see if the tests pass :)

---
In reply to: [451070402](https://github.com/dotnet/roslyn/pull/45684#discussion_r451070402) [](ancestors = 451070402)

333fred commented on src/Compilers/CSharp/Portable/Syntax/Syntax.xml:
```diff
@@ -237,18 +237,29 @@
+      </PropertyComment>
+    </Field>
+  </Node>
+  <Node Name="FunctionPointerParameterList" Base="CSharpSyntaxNode">
```
>FunctionPointerParameterList [](start = 14, length = 28)

Needs to be FunctionPointerParameterListSyntax

AlekseyTs commented on src/Compilers/CSharp/Portable/Syntax/Syntax.xml:
```diff
@@ -260,6 +271,59 @@
+      <summary>Function pointer calling convention syntax.</summary>
+    </TypeComment>
+    <Kind Name="FunctionPointerCallingConvention"/>
+    <Field Name="ManagedSpecifier" Type="SyntaxToken">
```
>ManagedSpecifier [](start = 17, length = 16)

It feels like the name implies one of the possible keywords. I think we should try to find more "neutral" name. "ManagedOrUnmanagedKeyword"? This would follow an existing pattern from "ThisOrBaseKeyword"

AlekseyTs commented on src/Compilers/CSharp/Portable/Syntax/Syntax.xml:
```diff
@@ -4052,6 +4132,23 @@
+      <summary>Parameter syntax.</summary>
+    </TypeComment>
+    <Kind Name="FunctionPointerParameter"/>
+    <Field Name="AttributeLists" Type="SyntaxList&lt;AttributeListSyntax&gt;" Override="true">
```
><Field Name="AttributeLists" Type="SyntaxList&lt;AttributeListSyntax&gt;" Override="true"> [](start = 4, length = 90)

It looks like this field is overridden the same way in both derived nodes. Is there a reason to do the overriding? Would simple inheritance work?

333fred replied:
Because the generator doesn't generate implementations of the properties in the abstract base class, it leaves them abstract.

---
In reply to: [451217460](https://github.com/dotnet/roslyn/pull/45684#discussion_r451217460) [](ancestors = 451217460)

AlekseyTs commented on src/Compilers/CSharp/Portable/Syntax/Syntax.xml:
```diff
@@ -4052,6 +4132,23 @@
+        <summary>Gets the attribute declaration list.</summary>
+      </PropertyComment>
+    </Field>
+    <Field Name="Modifiers" Type="SyntaxList&lt;SyntaxToken&gt;" Override="true">
```
><Field Name="Modifiers" Type="SyntaxList&lt;SyntaxToken&gt;" Override="true"> [](start = 4, length = 77)

Same question

AlekseyTs commented on src/Compilers/CSharp/Portable/Syntax/SyntaxKind.cs:
```diff
@@ -205,6 +205,8 @@ public enum SyntaxKind : ushort
         InitKeyword = 8443,
         RecordKeyword = 8444,
+        ManagedKeyword = 8445,
+        UnmanagedKeyword = 8446,
```
> UnmanagedKeyword = 8446, [](start = 7, length = 25)

Consider adding GetText method to the list in the comment below.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:
```diff
@@ -135,10 +136,11 @@ internal static class ParameterHelpers
 
                 if (mustBeLastParameter == null)
```
>if (mustBeLastParameter == null) [](start = 16, length = 32)

It looks like we can safely move this ```if``` statement right after the code that reports ERR_ThisInBadContext. This way all checks specific to ```ParameterSyntax``` will be grouped together and can be placed under umbrella of a single type test.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -98,6 +93,37 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+                    case { ManagedSpecifier: { RawKind: (int)SyntaxKind.ManagedKeyword } }:
+                        Debug.Assert(callingConventionSyntax.UnmanagedCallingConventionList is null || callingConventionSyntax.ContainsDiagnostics);
+                        return CallingConvention.Default;
+                    case { ManagedSpecifier: { RawKind: (int)SyntaxKind.UnmanagedKeyword }, UnmanagedCallingConventionList: { } specifiers } when isCallingConvention(specifiers, "Cdecl"):
```
>{ RawKind: (int)SyntaxKind.UnmanagedKeyword }, UnmanagedCallingConventionList: { } specifiers }  [](start = 45, length = 96)

Repeating this on every line making the code too verbose and hard to follow. It also create significant amount of code duplication. Please use nested switch instead, or something similar, and collapse duplicate logic.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -98,6 +93,37 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+                }
+
+                static bool isCallingConvention(FunctionPointerUnmanagedCallingConventionListSyntax specifiers, string expected)
+                    => specifiers.CallingConventions[0].Name.Text == expected;
```
>[0] [](start = 52, length = 3)

Is it always safe to index? Could the list be empty?

AlekseyTs replied:
>specifiers.CallingConventions[0].Name.Text == expected [](start = 23, length = 54)

For now, we probably also want to report an error when there is more than one identifier in the list.

AlekseyTs commented on src/Compilers/CSharp/Portable/GlobalSuppressions.cs:
```diff
@@ -34,7 +34,7 @@
 [assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters", Justification = "<Pending>", Scope = "member", Target = "~M:Microsoft.CodeAnalysis.CSharp.CSharpSyntaxTree.ParseText(Microsoft.CodeAnalysis.Text.SourceText,Microsoft.CodeAnalysis.CSharp.CSharpParseOptions,System.String,System.Collections.Immutable.Immutab ...
 [assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0027:Public API with optional parameter(s) should have the most parameters amongst its public overloads.", Justification = "<Pending>", Scope = "member", Target = "~M:Microsoft.CodeAnalysis.CSharp.SyntaxFactory.DeconstructionPatternClause(Microsoft.CodeAnalysis.SeparatedSyntaxList{Microsoft.CodeAnalysis.CSharp.Syntax.Subpa ...
-[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0027:Public API with optional parameter(s) should have the most parameters amongst its public overloads.", Justification = "<Pending>", Scope = "member", Target = "~M:Microsoft.CodeAnalysis.CSharp.SyntaxFactory.FunctionPointerType(Microsoft.CodeAnalysis.SeparatedSyntaxList{Microsoft.CodeAnalysis.CSharp.Syntax.ParameterSynt ...
 [assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters", Justification = "<Pending>", Scope = "member", Target = "~M:Microsoft.CodeAnalysis.CSharp.CSharpExtensions.GetDeclaredSymbol(Microsoft.CodeAnalysis.SemanticModel,Microsoft.CodeAnalysis.CSharp.Syntax.CompilationUnitSyntax,System.Threading.CancellationToken ...
```
It is not clear why this line is removed

333fred replied:
Because the API no longer exists.

---
In reply to: [451225860](https://github.com/dotnet/roslyn/pull/45684#discussion_r451225860) [](ancestors = 451225860)

AlekseyTs commented:
> 

It is not clear why this line is removed

---
Refers to: src/Compilers/CSharp/Portable/GlobalSuppressions.cs:41 in c4be137. [](commit_id = c4be1376a6e0465ef35f3dd44267659f5d1f104e, deletion_comment = True)

AlekseyTs commented on src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -211,6 +297,9 @@ static Microsoft.CodeAnalysis.CSharp.SyntaxFactory.WithExpression(Microsoft.Code
 *REMOVED*Microsoft.CodeAnalysis.CSharp.Syntax.ObjectCreationExpressionSyntax.NewKeyword.get -> Microsoft.CodeAnalysis.SyntaxToken
+*REMOVED*Microsoft.CodeAnalysis.CSharp.Syntax.ParameterSyntax.AttributeLists.get -> Microsoft.CodeAnalysis.SyntaxList<Microsoft.CodeAnalysis.CSharp.Syntax.AttributeListSyntax>
+*REMOVED*Microsoft.CodeAnalysis.CSharp.Syntax.ParameterSyntax.Modifiers.get -> Microsoft.CodeAnalysis.SyntaxTokenList
+*REMOVED*Microsoft.CodeAnalysis.CSharp.Syntax.ParameterSyntax.Type.get -> Microsoft.CodeAnalysis.CSharp.Syntax.TypeSyntax
```
The removal looks suspicious. Could this be a source of a binary break?

333fred replied:
We add an override instead (lines 128-130). Since we always prefer to emit `callvirt` instructions, it should be fine for most use cases. We did the same change when we introduced BaseObjectCreationExpressionSyntax as a new parent node.

---
In reply to: [451226984](https://github.com/dotnet/roslyn/pull/45684#discussion_r451226984) [](ancestors = 451226984)

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6393,9 +6397,54 @@ private ScanTypeFlags ScanFunctionPointerType(out SyntaxToken lastTokenOfType)
+                            lastTokenOfType = EatToken();
+                        }
+
+                        lastTokenOfType = EatToken(SyntaxKind.CloseBracketToken);
```
>EatToken(SyntaxKind.CloseBracketToken) [](start = 42, length = 38)

What happens here when there is no ```CloseBracketToken```? Are we going to return a fake token as last token? This doesn't feel right.

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6393,9 +6397,54 @@ private ScanTypeFlags ScanFunctionPointerType(out SyntaxToken lastTokenOfType)
+                    {
+                        while (true)
+                        {
+                            lastTokenOfType = EatToken(SyntaxKind.IdentifierToken);
```
>EatToken(SyntaxKind.IdentifierToken); [](start = 46, length = 37)

What happens here when there is no identifier? Are we going to return a fake token as last token? This doesn't feel right.

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6950,76 +6994,96 @@ private FunctionPointerTypeSyntax ParseFunctionPointerTypeSyntax()
+                    if (managedSpecifier.Kind == SyntaxKind.ManagedKeyword && unmanagedCallingConventions != null)
+                    {
+                        // 'managed' calling convention cannot be combined with unmanaged calling convention specifiers.
+                        unmanagedCallingConventions = AddError(unmanagedCallingConventions, ErrorCode.ERR_CannotSpecifyManagedWithUnmanagedSpecifiers);
```
>unmanagedCallingConventions [](start = 63, length = 27)

Consider attaching the first token rather than to the entire node. I think we have a helper to do that.

333fred replied:
I actually intentionally covered the entire node. The first token would just be the `[`, which I don't think would be great. I'd prefer to capture the entire thing.

---
In reply to: [451232665](https://github.com/dotnet/roslyn/pull/45684#discussion_r451232665) [](ancestors = 451232665)

AlekseyTs replied:
> I actually intentionally covered the entire node. The first token would just be the [, which I don't think would be great. I'd prefer to capture the entire thing.

Please consider that keeping smaller error span often leads to a better experience because there is less chances that spans of different errors are going to overlap.

---
In reply to: [451851393](https://github.com/dotnet/roslyn/pull/45684#discussion_r451851393) [](ancestors = 451851393,451232665)

AlekseyTs commented on src/Tools/Source/CompilerGeneratorTools/Source/VisualBasicSyntaxGenerator/GreenNodes/GreenNodeWriter.vb:
```diff
@@ -848,15 +848,15 @@ Friend Class GreenNodeWriter
         For i = 0 To allChildren.Count - 1
             If allChildren(i).IsList Then
-                _writer.WriteLine("            Dim {0} = VisitList(node.{1})" + vbCrLf +
+                _writer.WriteLine("            Dim {0} = VisitList(node.{1})" + Environment.NewLine +
```
>Environment.NewLine [](start = 80, length = 19)

It doesn't look like changes to VisualBasicSyntaxGenerator belong in this PR.

333fred replied:
This was from a separate PR, already merged into master. I'll rebase this out when review is complete, but it made my life simpler when running the generator.

---
In reply to: [451235615](https://github.com/dotnet/roslyn/pull/45684#discussion_r451235615) [](ancestors = 451235615)

AlekseyTs commented:
Done with review pass (iteration 8)

333fred commented:
> 

Because the API no longer exists.

---
In reply to: [655220662](https://github.com/dotnet/roslyn/pull/45684#issuecomment-655220662) [](ancestors = 655220662)

---
Refers to: src/Compilers/CSharp/Portable/GlobalSuppressions.cs:41 in c4be137. [](commit_id = c4be1376a6e0465ef35f3dd44267659f5d1f104e, deletion_comment = True)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [655231532](https://github.com/dotnet/roslyn/pull/45684#issuecomment-655231532) [](ancestors = 655231532)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -98,6 +93,68 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+                            case { CallingConventions: { Count: 1 } specifiers }:
+                                return specifiers[0].Name switch
+                                {
+                                    { Text: "Cdecl" } => CallingConvention.CDecl,
```
>Text [](start = 38, length = 4)

Should it be Text or ValueText? Throughout this switch. We probably should have tests with escaped identifiers.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -98,6 +93,68 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+                                // PROTOTYPE(func-ptr): Handle multiple specifiers
+                                foreach (var convention in specifiers)
+                                {
+                                    _ = reportBadConventionAndReturn(convention.Name.Text, convention.Name, diagnostics);
```
>reportBadConventionAndReturn(convention.Name.Text, convention.Name, diagnostics) [](start = 40, length = 80)

I guess this diagnostics could be misleading because the identifiers could be from the list the messages suggests as acceptable. Probably fine as a temporary solution until we get to addressing the PROTOTYPE comment

333fred replied:
Yeah, my intention is that much of this method will have to be rewritten, and that we'll only flag actually invalid things. I wanted to do the simplest possible thing for now.

---
In reply to: [452354819](https://github.com/dotnet/roslyn/pull/45684#discussion_r452354819) [](ancestors = 452354819)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -89,6 +92,35 @@ void symbolValidator(ModuleSymbol module)
+        }
+
+        [Fact]
+        public void UnrecognizedCallingConventions()
```
>UnrecognizedCallingConventions [](start = 20, length = 30)

It looks like this test is always going to pass.

333fred replied:
Whoops, meant to delete this entirely, I expanded the existing test.

---
In reply to: [452358536](https://github.com/dotnet/roslyn/pull/45684#discussion_r452358536) [](ancestors = 452358536)

AlekseyTs commented:
Done with review pass (iteration 10)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [656247690](https://github.com/dotnet/roslyn/pull/45684#issuecomment-656247690) [](ancestors = 656247690)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -98,6 +93,68 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+                                    { ValueText: "Thiscall" } => CallingConvention.ThisCall,
+                                    { ValueText: "Fastcall" } => CallingConvention.FastCall,
+                                    // PROTOTYPE(func-ptr): Handle unrecognized specifiers
+                                    var name => reportBadConventionAndReturn(name.Text, name, diagnostics)
```
>Text [](start = 82, length = 4)

ValueText?

333fred commented:
@dotnet/roslyn-compiler for a second review.

333fred commented:
@jasonmalinowski did you have any more comments?

jasonmalinowski commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:
```diff
@@ -956,15 +956,21 @@ public static bool IsDefaultExpressionContext(this SyntaxTree syntaxTree, int po
             }
 
-            return token.Parent is ParameterSyntax { Parent: FunctionPointerTypeSyntax _ };
+            return token switch
```
You _might_ be able to merge this and the previous switch into a single pattern patch...

333fred replied:
I didn't want to duplicate the `parent is` check. When we can use C# 9 `or` patterns this can be cleaned up :)

jasonmalinowski commented on src/Workspaces/CSharpTest/Formatting/FormattingTests_FunctionPointers.cs:
```diff
@@ -31,19 +30,90 @@ unsafe class C
+            await AssertFormatAsync(expected, content);
+        }
+
+        [Fact(Skip = "PROTOTYPE(func-ptr): Even though the rule is returning put a space, it's getting ignored, resulting in delegate*invalid [")]
```
Ditto: as long s it's formatting the stuff that it does understand, I guess that's good.

333fred commented:
@dotnet/roslyn-compiler for a second review please

RikkiGibson commented:
Taking a look
